tests/form-filling.js

```diff
@@ -0,0 +1,101 @@
+const helper = require('./helper');
+const genericFormPo = require('./page-objects/forms/generic-form.po');
+const answeredQuestions = [];
+
+const getPaths = (obj, prevKey, r = []) => {
+  const results = r;
+  let evalObj = obj;
+  if (Array.isArray(obj)) {
+    evalObj = obj[0];
+  }
+  Object.keys(evalObj).forEach((key) => {
+    let resultKey = '';
+    if(prevKey){
+      resultKey = `${prevKey}/${key}`;
+    } else {
+      resultKey = key;
+    }
+
+    if(key === 'inputs'){
+      return;
+    }
+    if(typeof evalObj[key] !== 'object'){
+      results.push({path: resultKey, value: evalObj[key]});
+    }else {
+      getPaths(evalObj[key], resultKey, results);
+    }
+  });
+  return results;
+};
+
+
+const getVisibleQuestions = async () => {
+  const answers = await element.all(by.css('.current input:not([disabled],.ignore)'));
+  const results = [];
+
+  for(let i = 0; i < answers.length; i++){
+    let name = await answers[i].getAttribute('data-name');
+    let cssAttribute = 'data-name';
+    if(!name) {
+      name = await answers[i].getAttribute('name');
+      cssAttribute = 'name';
+    }
+    const type = await answers[i].getAttribute('type');
+    if (results.filter(e => e.name === name && e.type === type).length === 0) {
+      results.push({ 
+        name: name,
+        type: type,
+        css: `input[${cssAttribute}="${name}"]`
+      });
+    }
+  }
+  return results;
+};
+
+const fillForm = async (reportFields,reportName) => {
+  let answers = await getVisibleQuestions();
+  const pathKeys = getPaths(reportFields);
+  answers.forEach(async (answer) => {
+    if (answeredQuestions.includes(answer.name)){
+      //already answered this question
+      return;
+    }
+    const answerVal = pathKeys.find(key => `/${reportName}/${key.path}` === answer.name);
+    if(!answerVal){
+      return;
+    }
+    switch(answer.type) {
+    case 'checkbox': {
+      const checks = answerVal.value.split();
+      checks.forEach((check) => {
+        helper.clickElementNative(element(by.css(`${answer.css}[value=${check}]`)));
+      });
+      break;
+    }
+    case 'radio':
+      await helper.clickElementNative(element(by.css(`${answer.css}[value="${answerVal.value}"]`)));
+      break;
+    case 'date': {
+      const css = `${answer.css} + div input`;
+      element(by.css(css)).sendKeys(answerVal.value);
+      break;
+    }
+    default:
+      element(by.css(answer.css)).sendKeys(answerVal.value);
+    }
```

Comment:
Switch statements make me jumpy (no pun intended). How about moving the functionality from the branches into functions. Then you could even make an object of those and map the types to functions kind of like  this:

```
const radio = () => { ... }
const date = () => { ... }

answerActions = { radio, date }

...

answerActions[answer.type]();
```

Normally a big benefit of this style is the ability to have generic functions that aren't tied to one specific use case (the switch statement) and the ability to then unit test each independently. We're probably not going to be unit testing this test code though, so that's debatable. I still find it to be more readable though, especially as the switch statement grows.